pkg/reconciler/namespace/namespace.go

```diff
@@ -0,0 +1,191 @@
+	"k8s.io/apimachinery/pkg/api/equality"
+	apierrs "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	kubelisters "k8s.io/client-go/listers/core/v1"
+	"knative.dev/pkg/controller"
+	"knative.dev/pkg/logging"
+)
+
+const (
+	whitelistLabel = "networking.internal.knative.dev/namespaceCert"
+)
+
+var wildcardName = names.WildcardCertName
+
+// Reconciler implements controller.Reconciler for Certificate resources.
+type Reconciler struct {
+	*reconciler.Base
+
+	// listers index properties about resources
+	nsLister            kubelisters.NamespaceLister
+	knCertificateLister listers.CertificateLister
+
+	configStore configStore
+}
+
+// Check that our Reconciler implements controller.Reconciler
+var _ controller.Reconciler = (*Reconciler)(nil)
+
+// Reconciler implements controller.Reconciler for Namespace resources.
+func (c *Reconciler) Reconcile(ctx context.Context, key string) error {
+	logger := logging.FromContext(ctx)
+	ctx = c.configStore.ToContext(ctx)
+
+	original, err := c.nsLister.Get(key)
+
+	if apierrs.IsNotFound(err) {
+		logger.Errorf("Namespace %s in work queue no longer exists %s", key, err)
+		return nil
+	} else if err != nil {
+		return err
+	}
+
+	// Don't modify the informers copy
+	newNamespace := original.DeepCopy()
+
+	whitelist, exists := newNamespace.Labels[whitelistLabel]
+	if !exists || whitelist == "false" {
+		logger.Infof("Skipping namespace certificate for namespace %s", newNamespace.Name)
+		return nil
+	}
+
+	err = c.reconcile(ctx, newNamespace)
+	if err != nil {
+		c.Recorder.Event(newNamespace, corev1.EventTypeWarning, "InternalError", err.Error())
+	}
+	return err
+}
+
+func (c *Reconciler) reconcile(ctx context.Context, ns *corev1.Namespace) error {
+	cfg := config.FromContext(ctx)
+
+	// Only create wildcard certs for the default domain
+	defaultDomain := cfg.Domain.LookupDomainForLabels(nil)
+	if defaultDomain == "" {
+		c.Logger.Infof("Skipping wildcard certificate creation for namespace %s. No default domain defined.", ns.Name)
+		return nil
+	}
+
+	domain, err := getWildcardDomain(cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	if err != nil {
+		c.Logger.Errorf("Failed to apply domain template %s to domain %s and namespace %s", cfg.Network.DomainTemplate, defaultDomain, ns.Name)
+	}
+
+	desiredCert := makeDesiredCertificate(ns, []string{domain})
+
+	certs, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).List(metav1.ListOptions{})
+	if err != nil {
+		return err
+	}
+	existingCert, err := getExistingCert(domain, certs)
+
+	if apierrs.IsNotFound(err) {
+		cert, err := c.ServingClientSet.NetworkingV1alpha1().Certificates(ns.Name).Create(desiredCert)
+		if err != nil {
+			c.Logger.Errorw("Failed to create namespace certificate", zap.Error(err))
+			c.Recorder.Eventf(ns, corev1.EventTypeWarning, "Creation Failed",
+				"Failed to create Knative certificate %s/%s: %v", ns.Name, desiredCert.ObjectMeta.Name, err)
+			return nil
```

Comment:
should we return `err` here?

Reply:
Good catch. I think we should definitely return `err` here.